tools/opensnoop.py

```diff
@@ -137,38 +178,54 @@
 
     return 0;
 };
+"""
 
-int trace_return(struct pt_regs *ctx)
+bpf_text_kprobe_body_openat2 = """
 {
+    struct val_t val = {};
     u64 id = bpf_get_current_pid_tgid();
-    struct val_t *valp;
-    struct data_t data = {};
+    u32 pid = id >> 32; // PID is higher part
+    u32 tid = id;       // Cast and get the lower part
+    u32 uid = bpf_get_current_uid_gid();
 
-    u64 tsp = bpf_ktime_get_ns();
+    PID_TID_FILTER
+    UID_FILTER
+    FLAGS_FILTER
 
-    valp = infotmp.lookup(&id);
-    if (valp == 0) {
-        // missed entry
+    if (container_should_be_filtered()) {
         return 0;
     }
-    bpf_probe_read_kernel(&data.comm, sizeof(data.comm), valp->comm);
-    bpf_probe_read_user(&data.fname, sizeof(data.fname), (void *)valp->fname);
-    data.id = valp->id;
-    data.ts = tsp / 1000;
-    data.uid = bpf_get_current_uid_gid();
-    data.flags = valp->flags; // EXTENDED_STRUCT_MEMBER
-    data.ret = PT_REGS_RC(ctx);
 
-    events.perf_submit(ctx, &data, sizeof(data));
-    infotmp.delete(&id);
+    if (bpf_get_current_comm(&val.comm, sizeof(val.comm)) == 0) {
+        val.id = id;
+        val.fname = filename;
+        val.flags = how->flags; // EXTENDED_STRUCT_MEMBER
```

Comment:
For this function, I think the main difference between it and generic trace_entry is the above line.
Could you add  something like
```
  int flags = how->flags;
```
in bpf_text_kprobe_header_openat2, it will eliminate this function, right?
something like
```
bpf_text_kprobe_header_openat2 = """
#include <uapi/linux/openat2.h>
int syscall__trace_entry_openat2(struct pt_regs *ctx, int dfd, const char __user *filename, struct open_how *how)
{
   int flags = how->flags;
"""
You need to adjust `syscall__trace_entry_open` and `syscall__trace_entry_openat` as well.

Reply:
Thanks. I'll update with it.